src/entry.c

```diff
@@ -0,0 +1,363 @@
+        return *entryGetValueRef(entry);
+    } else {
+        /* Skip field content, field null terminator and value sds8 hdr. */
+        size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
+        serverAssert((char *)entry + offset);
+
+        return (char *)entry + offset;
+    }
+}
+
+/* Modify the value of this entry and return a pointer to the (potentially new) entry.
+ * The value is taken by the function and cannot be reused after this function returns. */
+entry *entrySetValue(entry *e, sds value) {
+    if (entryHasValuePtr(e)) {
+        sds *value_ref = entryGetValueRef(e);
+        sdsfree(*value_ref);
+        *value_ref = value;
+        return e;
+    } else {
+        entry *new_entry = entryUpdate(e, value, entryGetExpiry(e));
+        return new_entry;
+    }
+}
+
+/* Returns the address of the entry allocation. */
+void *entryAllocPtr(const entry *entry) {
+    char *buf = sdsAllocPtr(entry);
+    if (entryHasValuePtr(entry)) buf -= sizeof(sds *);
+    if (entryHasExpiry(entry)) buf -= sizeof(long long);
+    return buf;
+}
+
+/**************************************** Entry Expiry API *****************************************/
+
+/* Returns the entry expiration timestamp.
+ * In case this entry has no expiration time, will return EXPIRE_NONE. */
+long long entryGetExpiry(const entry *entry) {
+    long long expiry = EXPIRY_NONE;
+    if (entryHasExpiry(entry)) {
+        char *buf = sdsAllocPtr(entry);
+        if (entryHasValuePtr(entry)) buf -= sizeof(sds *);
+        buf -= sizeof(expiry);
+        expiry = *(long long *)buf;
+    }
+    return expiry;
+}
+
+/* Modify the expiration time of this entry and return a pointer to the (potentially new) entry. */
+entry *entrySetExpiry(entry *e, long long expiry) {
+    if (entryHasExpiry(e)) {
+        char *buf = sdsAllocPtr(e);
+        if (entryHasValuePtr(e)) buf -= sizeof(sds *);
+        buf -= sizeof(expiry);
+        memcpy(buf, &expiry, sizeof(expiry));
+        return e;
+    }
+    entry *new_entry = entryUpdate(e, NULL, expiry);
+    return new_entry;
+}
+
+/* Return true in case the entry has assigned expiration or false otherwise. */
+int entryIsExpired(entry *entry) {
+    /* Don't expire anything while loading. It will be done later. */
+    if (server.loading) return 0;
+    if (!timestampIsExpired(entryGetExpiry(entry))) return 0;
+    if (server.primary_host == NULL && server.import_mode) {
+        if (server.current_client && server.current_client->flag.import_source) return 0;
+    }
+    return 1;
+}
+/**************************************** Entry Expiry API - End *****************************************/
+
+void entryFree(entry *entry) {
+    if (entryHasValuePtr(entry)) {
+        sdsfree(*entryGetValueRef(entry));
+    }
+    zfree(entryAllocPtr(entry));
+}
+
+/* Takes ownership of value. does not take ownership of field */
+entry *entryCreate(sds field, sds value, long long expiry) {
+    sds embedded_field_sds;
```

Comment:
In the interest of DRY, it feels like this function is doing 2 things:  1) allocating a properly sized block of memory, and 2) filling it in.  The "filling it in" part seems a lot like what `entryUpdate` is doing.  Is there room for simplification by leveraging `entryUpdate` to do a portion of the work, and removing repetition here?